groups/memory_group_service.go

```diff
@@ -0,0 +1,186 @@
+package groups
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/topfreegames/pitaya/config"
+	"github.com/topfreegames/pitaya/constants"
+)
+
+var (
+	memoryGroupsMu sync.RWMutex
+	memoryGroups   map[string]*MemoryGroup
+	memoryOnce     sync.Once
+)
+
+// MemoryGroupService base in server memory solution
+type MemoryGroupService struct {
+}
+
+// MemoryGroup is the struct stored in each group key(which is the name of the group)
+type MemoryGroup struct {
+	Uids        []string
+	LastRefresh int64
+	TTL         int64
+}
+
+// NewMemoryGroupService returns a new group instance
+func NewMemoryGroupService(conf *config.Config) *MemoryGroupService {
+	memoryOnce.Do(func() {
+		memoryGroups = make(map[string]*MemoryGroup)
+		go func() {
+			for now := range time.Tick(conf.GetDuration("pitaya.groups.memory.tickduration")) {
+				memoryGroupsMu.Lock()
+				for groupName, mg := range memoryGroups {
+					if mg.TTL != 0 && now.UnixNano()-mg.LastRefresh > mg.TTL {
+						delete(memoryGroups, groupName)
+					}
+				}
+				memoryGroupsMu.Unlock()
+			}
+		}()
+	})
+	return &MemoryGroupService{}
+}
+
+func getMemoryGroup(groupName string) (*MemoryGroup, error) {
+	memoryGroupsMu.Lock()
+	defer memoryGroupsMu.Unlock()
+
+	mg, ok := memoryGroups[groupName]
+	if !ok {
+		return nil, constants.ErrGroupNotFound
+	}
+
+	return mg, nil
+}
+
+// GroupCreate creates a group without TTL
+func (c *MemoryGroupService) GroupCreate(ctx context.Context, groupName string) error {
+	memoryGroupsMu.Lock()
+	defer memoryGroupsMu.Unlock()
+
+	if _, ok := memoryGroups[groupName]; ok {
+		return constants.ErrGroupAlreadyExists
+	}
+
+	memoryGroups[groupName] = &MemoryGroup{}
+	return nil
+}
+
+// GroupCreateWithTTL creates a group with TTL, which the go routine will clean later
+func (c *MemoryGroupService) GroupCreateWithTTL(ctx context.Context, groupName string, ttlTime time.Duration) error {
+	memoryGroupsMu.Lock()
+	defer memoryGroupsMu.Unlock()
+
+	if _, ok := memoryGroups[groupName]; ok {
+		return constants.ErrGroupAlreadyExists
+	}
+
+	memoryGroups[groupName] = &MemoryGroup{LastRefresh: time.Now().UnixNano(), TTL: ttlTime.Nanoseconds()}
+	return nil
+}
+
+// GroupMembers returns all member's UID in given group
+func (c *MemoryGroupService) GroupMembers(ctx context.Context, groupName string) ([]string, error) {
+	mg, err := getMemoryGroup(groupName)
```

Comment:
shouldn't you use the lock here?
also it seems that it would be more optimal if you have a group-specific mutex for changes like add and remove members and another one that is global for groupcreate and delete...

Reply:
Changed the implementations, to the point that every operation will be atomic. About the implementation of global lock and local lock, it would introduce a large complexity into the code, which can also cause more harm than good. Another solution would be to use sync.Map, but it would also be a global lock...